### ISSUE TITLE
Pipe conveys inconsistent value in GPU env

### ISSUE THREAD

**Issue by Daniel-Kant** [python/cpython#96251]
# Bug report

The Pipe and Queue of `multiprocessing` module conveys inconsistent value in GPU env while run correctly in non-heterogeneous env.
For the code below, the values on two sides of the Pipe are inconsistent. However, if the env is non-heterogeneous (do not `.to('cuda:0')`), the result is correct.
```python
import multiprocessing
import torch
import time

def genData(pipeIn):
    for i in range(2):
      t=torch.ones(3,).to('cuda:0')
      #t=torch.ones(3,)
      pipeIn.send(t)
    time.sleep(1)  


if __name__ == '__main__':
  pout, pin = multiprocessing.Pipe(duplex=False)
  temp = multiprocessing.Process(target=genData, args=(pin,))
  temp.start()
  for i in range(2):
    r = pout.recv()
    print(r)
    del r
```

# Environment

- CPython versions: 3.8.10
- Operating system and architecture: 
Windows 11 (Version: 21H2) on 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz   2.30 GHz
- GPU: NVIDIA GeForce RTX 3060 Laptop GPU 
- Dependecies: Torch (1.11.0+cu113)

**Comment by ronaldoussoren** [comment]
I just noticed that I did add the pending label, but never posted the comment that led me to do so.

I'm not a CUDA expert, but looking at the documentation the CUDA context (e.g. "cuda:0") is per process. Appearently there is a way to share contexts between processen, but that's not something enabled by default (for obvious reasons).

The PyTorch project may known how to accomplish what you are trying to accomplish.
